File: src/portrait_dataset_builder/taxonomy.py
"""Shared taxonomy for WHO? Portrait Dataset Builder.

Single source of truth for all classification labels used across
backend, API, frontend, and tests.
"""
from __future__ import annotations

# ── Horizontal pose (yaw-based) ──────────────────────────────────────────────

class HorizontalPose:
    FRONTAL = "frontal"
    QUARTER_LEFT = "three_quarter_left"
    QUARTER_RIGHT = "three_quarter_right"
    PROFILE_LEFT = "profile_left"
    PROFILE_RIGHT = "profile_right"

    ALL = [FRONTAL, QUARTER_LEFT, QUARTER_RIGHT, PROFILE_LEFT, PROFILE_RIGHT]

    # UI grouping: user-facing label → list of backend values
    GROUPS: dict[str, list[str]] = {
        "frontal": [FRONTAL],
        "quarter": [QUARTER_LEFT, QUARTER_RIGHT],
        "profile": [PROFILE_LEFT, PROFILE_RIGHT],
    }

    # User-friendly display labels
    LABELS: dict[str, str] = {
        FRONTAL: "Frontal",
        QUARTER_LEFT: "\u00be Left",
        QUARTER_RIGHT: "\u00be Right",
        PROFILE_LEFT: "Profile Left",
        PROFILE_RIGHT: "Profile Right",
    }


# ── Vertical pose (pitch-based) ──────────────────────────────────────────────

class VerticalPose:
    NEUTRAL = "neutral"
    LOOKING_UP = "looking_up"
    LOOKING_DOWN = "looking_down"

    ALL = [NEUTRAL, LOOKING_UP, LOOKING_DOWN]

    LABELS: dict[str, str] = {
        NEUTRAL: "Neutral",
        LOOKING_UP: "Looking Up",
        LOOKING_DOWN: "Looking Down",
    }


# ── Expression ───────────────────────────────────────────────────────────────

class ExpressionLabel:
    NEUTRAL = "neutral"
    SMILE = "smile"
    LAUGH = "laugh"
    SPEAKING = "speaking"

    ALL = [NEUTRAL, SMILE, LAUGH, SPEAKING]

    LABELS: dict[str, str] = {
        NEUTRAL: "Neutral",
        SMILE: "Smile",
        LAUGH: "Laugh",
        SPEAKING: "Speaking",
    }


# ── Lighting ─────────────────────────────────────────────────────────────────

class LightingLabel:
    DARK = "dark"
    BRIGHT = "bright"
    BALANCED = "balanced"

    ALL = [DARK, BRIGHT, BALANCED]

    LABELS: dict[str, str] = {
        DARK: "Dark",
        BRIGHT: "Bright",
        BALANCED: "Balanced",
    }


# ── Quality ──────────────────────────────────────────────────────────────────

class QualityLevel:
    HIGH = "high"
    MEDIUM = "medium"
    LOW = "low"

    ALL = [HIGH, MEDIUM, LOW]

    RANGES: dict[str, tuple[float, float]] = {
        HIGH: (0.80, 1.0),
        MEDIUM: (0.50, 0.79),
        LOW: (0.0, 0.49),
    }

    LABELS: dict[str, str] = {
        HIGH: "High (80+)",
        MEDIUM: "Medium (50-80)",
        LOW: "Low (<50)",
    }

File: who-app/src/types/taxonomy.ts
/**
 * Shared taxonomy for WHO? Portrait Dataset Builder.
 *
 * Single source of truth for all classification labels used across
 * frontend, API, and backend.
 */

export type HorizontalPose =
  | 'frontal'
  | 'three_quarter_left'
  | 'three_quarter_right'
  | 'profile_left'
  | 'profile_right';

export type VerticalPose = 'neutral' | 'looking_up' | 'looking_down';

export type ExpressionLabel = 'neutral' | 'smile' | 'laugh' | 'speaking';

export type LightingLabel = 'dark' | 'bright' | 'balanced';

export type QualityLevel = 'high' | 'medium' | 'low';

// UI grouping: user-facing filter key → backend values
export const HORIZONTAL_POSE_GROUPS: Record<string, HorizontalPose[]> = {
  frontal: ['frontal'],
  quarter: ['three_quarter_left', 'three_quarter_right'],
  profile: ['profile_left', 'profile_right'],
};

// User-friendly display labels
export const HORIZONTAL_POSE_LABELS: Record<HorizontalPose, string> = {
  frontal: 'Frontal',
  three_quarter_left: '\u00be Left',
  three_quarter_right: '\u00be Right',
  profile_left: 'Profile Left',
  profile_right: 'Profile Right',
};

export const VERTICAL_POSE_LABELS: Record<VerticalPose, string> = {
  neutral: 'Neutral',
  looking_up: 'Looking Up',
  looking_down: 'Looking Down',
};

export const EXPRESSION_LABELS: Record<ExpressionLabel, string> = {
  neutral: 'Neutral',
  smile: 'Smile',
  laugh: 'Laugh',
  speaking: 'Speaking',
};

export const LIGHTING_LABELS: Record<LightingLabel, string> = {
  dark: 'Dark',
  bright: 'Bright',
  balanced: 'Balanced',
};

export const QUALITY_LABELS: Record<QualityLevel, string> = {
  high: 'High (80+)',
  medium: 'Medium (50-80)',
  low: 'Low (<50)',
};

export const QUALITY_RANGES: Record<QualityLevel, [number, number]> = {
  high: [0.80, 1.0],
  medium: [0.50, 0.79],
  low: [0.0, 0.49],
};
